### ISSUE TITLE
strange case that calibration does not work

### ISSUE THREAD

**Issue by ShellingFord221** [kundajelab/labelshiftexperiments#4]
Recently I conduct an experiment of classification under domain shift, but the evaluation metric is F1 score instead of accuracy. In my settings, the label proportion in test set is known, therefore I just need to optimize parameters of BCTS on development set instead of performing EM (I still use cross-entropy as training loss). But the results are very strange. The best parameters (saved when accuracy achieves best or F1 on dev set achieves best) leads to a very low F1 on test set (e.g. 0.08), pretty much lower than that of an un-calibrated model (e.g. 0.3).
Is optimization of accuracy not suitable for F1 evaluation, or is there something wrong with my implementation? Thanks a lot!

**Comment by AvantiShri** [comment]
Hi @ShellingFord221, I am confused by the statement “I just need to optimize parameters of BCTS on development set instead of performing EM”. BCTS just does calibration; it does not do adaptation to label shift. If you know the class proportion in the test set then you don’t need to do EM but you *do* need to apply Bayes’ rule (the last step of EM) to adapt the predictions.

**Comment by AvantiShri** [comment]
In terms of the code, you can use the class PriorShiftAdapterFunc (https://github.com/blindauth/abstention/blob/b9585af9ed54b70feea102ad63f5ab6f0f6285e0/abstention/label_shift.py#L14)

You would set the `multipliers` argument in `__init__` to be the ratio q(y)/p(y), where q(y) is the class proportions on the test set and p(y) is the class proportions on the training set. You should then call this object on the calibrated predictions produced using BCTS (you don't have to set the `calibrator_func` argument in `__init__` if you are calling it on probabilities that have already been calibrated).

**Comment by ShellingFord221** [comment]
Thanks for the details! I think the best way to tackle label shift with known distribution in test set is to first calibrate the original classifier on dev set (i.e. optimize parameters of BCTS on dev set), then using Eq. 8 (i.e. multiply lable ratio) to make calibrated model adapt to label shift. In my experiments, F1 score of calibrated model is better than the original model (they both don't use Eq. 8), but F1 score of a calibrated model using Eq. 8 is worse than that of an _uncalibrated_ model using Eq. 8 (i.e. just multiplying the ratio q(y)/p(y)). It seems that calibration is not necessarily needed when label shift adaptation is adopted. We talk about this in Issue #2, the intuition of calibration is quite clear (although experiments may not support very well ), could you please give some references of the necessarity of calibration under label shift? Thank you so much!

**Comment by AvantiShri** [comment]
Hi @ShellingFord221 are you able to send me the predictions+labels on the validation and testing sets so that I can look closer at what may be going on?

**Comment by ShellingFord221** [comment]
Hi, sorry for the late reply! I conduct my experiments on a NLP task, which is designed to prove that calibration is necessary under label shift. Adaption only will hurt the performance. But after re-reading your paper, I think step 2 may be ignored. I just use label distribution in training set to represent p_hat(y=i) in Eq. 8, which is actually the average of predictions over dev set. I think this is the place where calibration works: to better estimate source-domain priors. As for EM, it is used to estimate target-domain distribution given source-domain priors. If there is bias in the model, the prior will be poorly-estimated, then the target-domain distribution will be poorly-estimated, thus the performance will be hurt. If this is correct, then here comes my question: why can't we just count samples of each label in training set as the prior distribution (then we may not need to do calibration)? Or can we just count labels in dev set when calibration's bias is corrected? Thanks again for your kindly help!

**Comment by AvantiShri** [comment]
Hi @ShellingFord221, calibration is not needed for the purpose of getting reliable estimates of p_hat(y=i) (as you noted, you can simply count the proportions of samples in the training set in order to get this). Rather, calibration is needed in order to get reliable estimates of p_hat(y=i|x). When calibration is bias-corrected, you are right that the average of the labels in the validation set (which I think you are calling the "dev set") will typically be equivalent to averaging p_hat(y=i|x) over the validation set after calibration. However, calibration is still needed in order to get the p_hat(y=i|x) itself (you need access to p_hat(y=i|x) in order to do the label shift adaptation).

There is an additional nuance, which I think you are getting at as well when you say "can we just count labels in dev set when calibration's bias is corrected?": if the bias-correction is not perfect (even when using BCTS, there may still be some small bias left), then there may still be a difference between the average of p_hat(y=i|x) over the validation set and the average of the labels. This is discussed in section 3.2 of the paper ("Defining source-domain priors in the EM"). In this case, you should use one of the following two approaches:
- Approach 1 (if you know the ground-truth label proportions for the target domain, which seems to be true for you):
(a) compute your estimates of p_hat(y=i) and q_hat(y=i) by averaging the ground-truth label proportions for the source and target domain respectively
(b) apply calibration using the validation set to get reliable estimates of p_hat(y=i|x)
(c) apply equation 8 (with calibrated p_hat(y=i|x)) in order to get q_hat(y=i|x).
This skips both step 2 and step 3 of the algorithm, and I think it's what you are suggesting. But note that you still need calibration to get reliable estimates of p_hat(y=i|x).
- Approach 2 (if you do not know the ground truth label proportions for the target domain): use the standard approach outlined in the paper, where you compute p_hat(y=i) by averaging the *predictions* of the calibrated p_hat(y=i|x) over the validation set and estimate q_hat(y=i) via maximum likelihood. In this situation, you should definitely NOT estimate p_hat(y=i) by averaging the labels on the validation set for the reason outlined in section 3.2.

If your bias correction happened to be imperfect, and you were using the average of the predictions in order to estimate p_hat(y=i) but you were using the true label distribution to get q_hat(y=i), then that could definitely explain some of your strange results. However, it would be easy to check this - do the results of averaging the calibrated predictions on the validation set look substantially different from the true label distribution?

If you send me a file that has the predictions and labels for the validation and test sets, I may be able to better diagnose the problem. No need to send me the input data or any other information - I would just need to look at the predictions and labels.

**Comment by ShellingFord221** [comment]
Hi, here are the label proportions of train_set (i.e. 80% of train.txt), dev_set (i.e. 20% of train.txt), test_set (i.e. test.txt) and predictions of a calibrated model on dev_set (there are 7 labels in total):

train_set: 0.7423, 0.0681, 0.0796, 0.0466, 0.0466, 0.0168, 0.0001 (very close to 0)
dev_set: 0.7503, 0.0599, 0.0842, 0.0452, 0.0454, 0.0150, 0.0000
dev_set (ave(prob)): 0.7522, 0.0840, 0.0654, 0.0446, 0.0398, 0.0140, 0.0000
test_set: 0.8826, 0.0439, 0.0177, 0.0080, 0.0139, 0.0005, 0.0332

Thank you so much!

**Comment by ShellingFord221** [comment]
Hi, here are the label proportions of train_set (i.e. 80% of train.txt), dev_set (i.e. 20% of train.txt), test_set (i.e. test.txt) and predictions of a calibrated model on dev_set and results of a calibrated model with adaptation on test_set (PS: there are 7 labels in total; calibration method is BCTS):

train_set: 0.7423, 0.0681, 0.0796, 0.0466, 0.0466, 0.0168, 0.0001 (very close to 0)
dev_set: 0.7503, 0.0599, 0.0842, 0.0452, 0.0454, 0.0150, 0.0000
dev_set (ave(prob)): 0.7522, 0.0840, 0.0654, 0.0446, 0.0398, 0.0140, 0.0000
test_set: 0.8826, 0.0439, 0.0177, 0.0080, 0.0139, 0.0005, 0.0332
test_set (ave(prob)): 0.7102, 0.0753, 0.0189, 0.0082, 0.0161, 0.0008, 0.1705

The performance is given as follows (evaluated by F1 score; label distribution in test set is totally known):
original: 0.2827
original + calibration: 0.3345
original + label adaptation: 0.3385
original + calibration + label adaptation: 0.2432
I have also tried another method, which does adaptation explicitly by set model bias as train_label_propotion in training and test_label_propotion in test. F1 score of this method is 0.3950.

The detailed files have been sent to you. Thank you so much!

**Comment by ShellingFord221** [comment]
Hi, here are the label proportions of train_set (i.e. 80% of train.txt), dev_set (i.e. 20% of train.txt), test_set (i.e. test.txt) and predictions of a calibrated model on dev_set and results of a calibrated model with adaptation on test_set (PS: there are 7 labels in total; calibration method is BCTS):

train_set: 0.7423, 0.0681, 0.0796, 0.0466, 0.0466, 0.0168, 0.0001 (very close to 0)
dev_set: 0.7503, 0.0599, 0.0842, 0.0452, 0.0454, 0.0150, 0.0000
dev_set (ave(prob)): 0.7522, 0.0840, 0.0654, 0.0446, 0.0398, 0.0140, 0.0000
test_set: 0.8826, 0.0439, 0.0177, 0.0080, 0.0139, 0.0005, 0.0332
test_set (ave(prob)): 0.7102, 0.0753, 0.0189, 0.0082, 0.0161, 0.0008, 0.1705

The performance is given as follows (evaluated by F1 score; label distribution in test set is totally known):
original: 0.2827
original + calibration: 0.3345
original + label adaptation: 0.3385
original + calibration + label adaptation: 0.2432
I have also tried another method, which does adaptation explicitly by set model bias as train_label_propotion in training and test_label_propotion in test. F1 score of this method is 0.3950.

The detailed files have been sent to you. Thank you so much!
